### PR TITLE
Add `sessionStorageId`

### LINKED ISSUE #157: Add `session_storage_id`
When the SDK is initialized, get or create an unique `session_storage_id` that is sent along the websocket query param.

@classicus-eth FYI I'm adding a `session_storage_id` to the EventV2 model

### PR DESCRIPTION
**Pull request by gtupak**
Added an unique `sessionStorageId` to the websocket connection string.

Closes #157

### CODE REVIEW

File: src/ArcxAnalyticsSdk.ts [hunk 1]
```diff
@@ -380,19 +389,31 @@ export class ArcxAnalyticsSdk {
...
+  private static _getSessionId() {
+    const existingSessionId = window.sessionStorage.getItem(SESSION_STORAGE_ID_KEY)
+    if (existingSessionId) {
+      return existingSessionId
+    }
+
+    const newSessionId = generateUniqueID()
+    window.sessionStorage.setItem(SESSION_STORAGE_ID_KEY, newSessionId)
+    return newSessionId
+  }
```
Comment (classicus-eth) [inline src/ArcxAnalyticsSdk.ts:415]: @tanghyd We talked about the storage yesterday, it actually CAN be persisted between refreshes and page visits:
https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage
> - A page session lasts as long as the tab or the browser is open, and survives over page reloads and restores.

File: src/utils/generateUniqueID.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+let counter = 0
+
+/**
+ * Generates a unique ID.
+ */
+export function generateUniqueID() {
+  const timestamp = new Date().getTime()
+  const randomPart = Math.random().toString(36).substring(2)
+  // Counter increments on every call to ensure uniqueness even if called rapidly in succession
+  counter += 1
```
Comment (classicus-eth) [inline src/utils/generateUniqueID.ts:10]: Curious why the counter is needed? 

If it's called in rapid succession, then the timestamp would be slightly different. 

If it's called by two different users at the same time (e.g. two different browsers) then the `Math.random()` would need to _also_ collide. Question is: the source of "randomness" in `Math.random()`, is it also time-based? 

Interesting read on `Math.random`: https://stackoverflow.com/questions/28199100/probability-of-getting-the-same-value-using-math-random/28220928#28220928 

^^ Probably overkill, but once it peaked my curiosity, I couldn't stop myself 🙃 

[Not a blocker]

**Review by classicus-eth — COMMENTED** [review]
Not explicitly defined timeout; if a user keeps a tab open forever, they will have an extremely long session duration (e.g. weeks).

Assuming we don't reset `sessionsStorageId` upon refresh, is this expected behavior?

File: src/utils/generateUniqueID.ts (hunk 2 above)
Comment (gtupak) [reply src/utils/generateUniqueID.ts:10]: NGL, I wrote this function with ChatGPT. Here's his answer:

---
You've asked a good question. The counter is indeed included to handle rapid successive calls that might happen within the same millisecond, where the timestamp value would be identical. Though JavaScript engines are not generally fast enough to execute the function multiple times within a single millisecond in usual circumstances, including a counter is a safeguard to maintain uniqueness in exceptional cases, adding an extra layer of uniqueness.

Regarding the Math.random() function, it is not time-based; it uses a pseudo-random number generator. The exact algorithm can vary between JavaScript engines, but it is generally not based directly on the current time. Common algorithms include things like the Xorshift or Mersenne Twister algorithms, which generate sequences of numbers that are statistically random, but if you know the current state of the generator, you can predict future outputs (hence "pseudo-random").

Even though Math.random() isn't time-based, the fact that it is seeded with some initial randomness (often at the time the JavaScript engine starts up) means that in practice, it's extremely unlikely (though not impossible) for two users to get the same "random" sequence, and hence for two identifiers generated at the same time to collide.

To improve the randomness, you might incorporate other sources of entropy, such as user events or data. But for many applications, a timestamp combined with Math.random() provides a sufficient level of uniqueness and randomness for generating identifiers.

---

I've added it for one additional layer of ensuring uniqueness but I agree it's overkill :P

**Comment by gtupak** [comment]
Not having a timeout was the intended behaviour because we agreed that the session calculation is done at the pipeline level, where the timeout will be enforced. That was my understanding from yesterday's call.

File: src/utils/generateUniqueID.ts (hunk 2 above)
Comment (classicus-eth) [reply src/utils/generateUniqueID.ts:10]: Could we add the identityId to the sessionId? 🤔 That would add the missing entropy pieces + would be beneficial in the data bakehouse in matching sessions with identities from a human-readably standpoint. Thats actually what Daniel did in the lakehouse to create sessionIds downstream (`{identityId}--1`).

That would for-sure guarantee uniqueness (or almost), as the counter only protects againts uniqueness within a single browser, not across browsers (e.g. two different users can go through `counter=1, counter=2,...` at the same timestamp).

Not sure if you can have an `storage_session_id` without having an `identity_id` tho. But from my knowledge they are both created at initialization of SDK if API key is valid, and we won't accept events if not anyway.

Thoughs?

```suggestion
  return `${timestamp}-${identityId}-${randomPart}`
```

Comment (gtupak) [reply src/utils/generateUniqueID.ts]: I like that idea! I pushed new code, check it out!

**Review by classicus-eth — APPROVED** [review]
☕

**Pull request merged**